pkg/controller/perconaservermongodb/psmdb_controller.go

```diff
@@ -291,7 +287,15 @@ func (r *ReconcilePerconaServerMongoDB) Reconcile(request reconcile.Request) (re
 			cr.Status.Replsets[replset.Name] = &api.ReplsetStatus{}
 		}
 
-		err = r.reconcileCluster(cr, replset, *pods, secrets)
+		isClusterLive, err := r.reconcileCluster(cr, replset, *pods, secrets)
+
+		defer func() {
+			err = r.updateStatus(cr, err, isClusterLive)
+			if err != nil {
+				reqLogger.Error(err, "failed to update cluster status", "replset", replset.Name)
+			}
+		}()
```

Comment:
Please move "defer" back to the previous place. Otherwise, it won't catch errors between lines 136 and 292.
Just create `isClusterLive := false` before the defer statement. So by default, we assume the cluster isn't alive until `reconcileCluster()` proves opposite. Furthermore `updateStatus` won't even reach the `isClusterLive` check if there are some errors so we're safe to have false there anyway.